### PR TITLE
Add hiding KZG polynomial commitment scheme

### PR DESCRIPTION
**Pull request by waamm**
## Description
- Adds the hiding KZG polynomial commitment scheme as described on [this](https://alinush.github.io/dekart) page. This also needed a barycentric evaluation function.
- Also does some small renaming, e.g. `fixedbasemsms` to `fixed_base_msms`.
- Refactors the tests in `tests/range_proof` to only generate prover keys and verifier keys once, and removes large test sizes, **to improve test speed**.

## How Has This Been Tested?
- There is a test inside of `hiding_kzg`. In the future it'll probably be generalised further as part of a trait.
- There is also a test for the new barycentric evaluation function inside of `tests/barycentric_eval`.

## Key Areas to Review
For security purposes it is important that `verify()` is correct, and that the refactored range proof tests are correct.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [x] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [x] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Move Compiler
- [x] Other: this code is not yet in use

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: crates/aptos-dkg/benches/range_proof.rs [hunk 1]
```diff
@@ -2,7 +2,11 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use aptos_dkg::{
-    range_proofs::{dekart_univariate::Proof as UnivariateDeKART, traits::BatchedRangeProof},
+    range_proofs::{
+        dekart_univariate::Proof as UnivariateDeKART,
+        //dekart_univariate_v2::Proof as UnivariateDeKARTv2,
```
Comment (alinush) [inline crates/aptos-dkg/benches/range_proof.rs:7]: What is happening here? Why are you introducing code that is commented out?

Comment (waamm) [reply crates/aptos-dkg/benches/range_proof.rs:7]: This was the result of moving `dekart_univariate_v2` to the next PR. If you prefer I can remove it and then readd it in the next PR.

Comment (alinush) [reply crates/aptos-dkg/benches/range_proof.rs:7]: Ah, it would be clearer to avoid stuff like this in the future (since it adds cognitive overhead during the review; that's why it's better to build the PR incrementally and open for review when it exceeds 500 lines).

But leave as is for now to save you some extra work 👌

**Comment by alinush** [comment]
@rex1fernando can you please also review this as soon as possible? 🙏

File: crates/aptos-dkg/src/algebra/polynomials.rs [hunk 2]
```diff
@@ -24,6 +24,96 @@ pub(crate) fn differentiate_in_place<F: Field>(coeffs: &mut Vec<F>) {
...
+///
+/// # Arguments
+/// * `f_vals` - Slice of field elements representing `f_i`.
+/// * `x_vals` - Slice of field elements representing `x_i`. Must have the same length as `f_vals`.
+/// * `x` - The field element `x` appearing in each denominator `(x_i - x)`.
+/// * `y` - The field element `y` subtracted from each `f_i` in the numerator.
+///
+/// # Returns
+/// A `Vec<F>` containing the evaluations of `(f_i - y) / (x_i - x)` for each index `i`.
+/// Or will panic if one of the `x_i` equals `x`.
```
Comment (alinush) [inline crates/aptos-dkg/src/algebra/polynomials.rs:42]: Would be helpful to add:
```
///
/// Useful to compute the scalars needed to interpolate a KZG proof (i.e., a quotient commitment) for $f(\gamma) = y$ as 
///    $\pi = ( f(\omega^i) - f(\gamma) ) / (\omega^i - \gamma) * L_i$
/// Here, $y_i = f(\omega^i), x_i$ = \omega^i, y=f(\gamma) and $x=\gamma$.
```

File: crates/aptos-dkg/src/algebra/polynomials.rs [hunk 3]
```diff
@@ -24,6 +24,96 @@ pub(crate) fn differentiate_in_place<F: Field>(coeffs: &mut Vec<F>) {
...
+    x: F,
+    y: F,
+) -> Vec<F> {
+    assert_eq!(f_vals.len(), x_vals.len());
+
+    // Step 1: compute denominators x_i - x
+    let mut denoms: Vec<F> = x_vals.iter().map(|&xi| xi - x).collect();
+
+    // Step 2: batch inversion in place
+    ark_ff::batch_inversion(&mut denoms);
```
Comment (alinush) [inline crates/aptos-dkg/src/algebra/polynomials.rs:55]: Nice! 👌 👌

File: crates/aptos-dkg/src/algebra/polynomials.rs [hunk 4]
```diff
@@ -24,6 +24,96 @@ pub(crate) fn differentiate_in_place<F: Field>(coeffs: &mut Vec<F>) {
...
+
+    // Step 3: multiply numerators by inverses
+    f_vals
+        .iter()
+        .zip(denoms.iter())
+        .map(|(&f_val, &denom_inv)| (f_val - y) * denom_inv)
+        .collect()
+}
+
+/// Evaluate a polynomial given by its values at roots of unity
```
Comment (alinush) [inline crates/aptos-dkg/src/algebra/polynomials.rs]: Editorial:
 - "a polynomial given by its values at roots of unity" --> "a polynomial given in Lagrange basis"

File: crates/aptos-dkg/src/algebra/polynomials.rs [hunk 5]
```diff
@@ -24,6 +24,96 @@ pub(crate) fn differentiate_in_place<F: Field>(coeffs: &mut Vec<F>) {
...
+    evals: &[F],
+    roots_of_unity_in_eval_dom: &[F],
+    x: F,
+    n_inv: F,
+) -> F {
+    let n = evals.len();
+    assert_eq!(n, roots_of_unity_in_eval_dom.len());
+
+    // If x exactly matches a root, return the corresponding evaluation
+    for (root, val) in roots_of_unity_in_eval_dom.iter().zip(evals.iter()) {
```
Comment (alinush) [inline crates/aptos-dkg/src/algebra/polynomials.rs]: This will slow us down for really large polynomials. And I presume it is never meaningfully used, right?

Maybe turn it into a `debug_assert!`? Then, in release mode, if this ever gets called with $x = \omega^i$ it will panic due to a division by zero somewhere.

File: crates/aptos-dkg/src/algebra/polynomials.rs [hunk 6]
```diff
@@ -24,6 +24,96 @@ pub(crate) fn differentiate_in_place<F: Field>(coeffs: &mut Vec<F>) {
...
+        if *root == x {
+            return *val;
+        }
+    }
+
+    // Compute denominators (x - omega^j)
+    let mut denoms: Vec<F> = roots_of_unity_in_eval_dom
```
Comment (alinush) [inline crates/aptos-dkg/src/algebra/polynomials.rs]: Alternatively, instead of the `debug_assert!`, you could return early here when you detect `x == omega_j`.

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+use ark_std::rand::{CryptoRng, RngCore};
+use sigma_protocol::homomorphism::TrivialShape as CodomainShape;
+
+#[derive(CanonicalSerialize, CanonicalDeserialize, Debug, Clone)]
+pub struct Commitment<E: Pairing>(pub E::G1);
+
+#[derive(CanonicalSerialize, CanonicalDeserialize, Debug, Clone)]
+pub struct CommitmentRandomness<E: Pairing>(pub E::ScalarField);
+
+// TODO: Rename to OpeningProof?
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs]: Yes!

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+        xi_1: ck.xi_1,
+    };
+
+    let input = (r.0, values.to_vec());
+
+    Commitment(commitment_hom.apply(&input).0)
+}
+
+impl<'a, E: Pairing> Homomorphism<'a, E> {
+    // TODO: should maybe make `y` part of the input, since it's often computed before invoking `open()`
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs]: Yes!

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+        let y =
+            polynomials::barycentric_eval(&f_evals, &ck.roots_of_unity_in_eval_dom, x, ck.m_inv);
+
+        let q_evals =
+            polynomials::quotient_evaluations_batch(&f_evals, &ck.roots_of_unity_in_eval_dom, x, y);
+
+        let pi_1 = commit_with_randomness(ck, &q_evals, s);
+
+        let pi_2 = (ck.one_1 * rho) - (ck.tau_1 - ck.one_1 * x) * s.0;
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:155]: Should be an MSM? Or is it slower in arkworks for this size?

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+    pub lagr_g1: &'a [E::G1Affine],
+    pub xi_1: E::G1Affine,
+}
+
+impl<'a, E: Pairing> homomorphism::Trait for Homomorphism<'a, E> {
+    type Codomain = CodomainShape<E::G1>;
+    type Domain = (E::ScalarField, Vec<E::ScalarField>);
+
+    fn apply(&self, input: &Self::Domain) -> Self::Codomain {
+        self.apply_msm(self.msm_terms(input))
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:200]: 🫣😆

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_ec::pairing::Pairing;
+    use ark_poly::{univariate::DensePolynomial, Polynomial};
+    use ark_std::{rand::thread_rng, UniformRand};
+
+    // TODO: Should set up a PCS trait, then make these tests generic?
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:245]: Eventually!

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+        // Polynomial values at the roots of unity
+        let f_evals: Vec<Fr<E>> = ck
+            .roots_of_unity_in_eval_dom
+            .iter()
+            .map(|&gamma| poly.evaluate(&gamma))
+            .collect();
+
+        let rho = CommitmentRandomness::<E>(Fr::<E>::rand(&mut rng));
+        let s = CommitmentRandomness::<E>(Fr::<E>::rand(&mut rng));
+        let x = Fr::<E>::rand(&mut rng);
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:269]: Nice!

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+        let y =
+            polynomials::barycentric_eval(&f_evals, &ck.roots_of_unity_in_eval_dom, x, ck.m_inv);
+
+        // Commit to f
+        let comm = super::commit_with_randomness(&ck, &f_evals, &rho);
+
+        // Open at x
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs]: This will fail when x is a root of unity.

Negligible in our large prime field so we are good 👌

File: crates/aptos-dkg/src/utils/test_utils.rs [hunk 14]
```diff
@@ -5,30 +5,22 @@ use crate::range_proofs::traits::BatchedRangeProof;
...
-    B::ProverKey,
-    B::VerificationKey,
-    Vec<B::Input>,
-    B::Commitment,
-    B::CommitmentRandomness,
-) {
-    let (pk, vk) = B::setup(n, ell, rng); // TODO: potentially change these values back to n + 10 and ell + 10?
-
+) -> (Vec<B::Input>, B::Commitment, B::CommitmentRandomness) {
+    // TODO: One might want to assert something like n <= pk.max_n here, for which you'd have to e.g. add a trait HasMaxN to ProverKey
```
Comment (alinush) [inline crates/aptos-dkg/src/utils/test_utils.rs:15]: Or the Sized trait would be more standard?

File: crates/aptos-dkg/tests/barycentric_eval.rs [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    let poly_coeffs: Vec<F> = (0..=degree).map(|_| F::rand(&mut rng)).collect();
+    let poly = DensePolynomial::from_coefficients_vec(poly_coeffs);
+
+    let n = 128;
+    let omega = F::get_root_of_unity(n).unwrap();
+    let roots: Vec<F> = (0..n).map(|i| omega.pow([i])).collect();
+    let evals: Vec<F> = roots.iter().map(|&root| poly.evaluate(&root)).collect();
+    let n_inv = F::from(n).inverse().unwrap();
+
+    for _ in 0..20 {
```
Comment (alinush) [inline crates/aptos-dkg/tests/barycentric_eval.rs]: I would lower this to like 5 and iterate through more possibilities for the degree?

Test degree 0 polynomials, including the zero polynomial, degree 1, 2 and, idk, 16.

File: crates/aptos-dkg/tests/barycentric_eval.rs [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+#[test]
+fn test_barycentric_eval() {
+    type F = ark_bls12_381::Fr;
+    let mut rng = ark_std::rand::thread_rng();
+
+    let degree = 64;
+    let poly_coeffs: Vec<F> = (0..=degree).map(|_| F::rand(&mut rng)).collect();
+    let poly = DensePolynomial::from_coefficients_vec(poly_coeffs);
+
+    let n = 128;
```
Comment (alinush) [inline crates/aptos-dkg/tests/barycentric_eval.rs]: You should also test n = degree + 1

File: crates/aptos-dkg/tests/range_proof.rs [hunk 17]
```diff
@@ -4,42 +4,34 @@
...
     utils::test_utils,
 };
+use ark_bls12_381::Bls12_381;
+use ark_bn254::Bn254;
 use ark_ec::pairing::Pairing;
 use ark_std::rand::thread_rng;
 
 #[cfg(test)]
-fn run_range_proof_completeness<E: Pairing, B: BatchedRangeProof<E>>(n: usize, ell: usize) {
+fn run_range_proof_roundtrip<E: Pairing, B: BatchedRangeProof<E>>(
```
Comment (alinush) [inline crates/aptos-dkg/tests/range_proof.rs]: I think this should be called assert_range_proof_correctness

File: crates/aptos-dkg/src/utils/test_utils.rs (hunk 14 above)
Comment (waamm) [reply crates/aptos-dkg/src/utils/test_utils.rs:15]: Hmm I thought `Sized` was a trait mostly intended to inform the compiler how much memory to allocate.

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs (hunk 9 above)
Comment (waamm) [reply crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:155]: Just checked and for BLS12-381 the difference is only 10% (in favour of the direct approach), whereas for BN254 the direct approach is even faster (110 µs vs 155 µs).

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+            (tau_2 - one_2 * x).into_affine(),
+            xi_2,
+        ]);
+        ensure!(PairingOutput::<E>::ZERO == check);
+
+        Ok(())
+    }
+}
+
+pub struct Homomorphism<'a, E: Pairing> {
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:190]: Why is this called Homomorphism? I don't understand. Let's discuss!

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_ec::pairing::Pairing;
+    use ark_poly::{univariate::DensePolynomial, Polynomial};
+    use ark_std::{rand::thread_rng, UniformRand};
+
+    // TODO: Should set up a PCS trait, then make these tests generic?
+    fn generic_test_open_and_verify_roundtrip<E: Pairing>() {
```
Comment (alinush) [inline crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs]: This should be called assert_kzg_opening_correctness

File: crates/aptos-dkg/tests/barycentric_eval.rs [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+        vec![Fr::<E>::ZERO]
+    } else {
+        (0..=degree).map(|_| Fr::<E>::rand(&mut rng)).collect()
+    };
+    let poly = DensePolynomial::from_coefficients_vec(poly_coeffs);
+
+    // Build domain and evaluations
+    let omega = Fr::<E>::get_root_of_unity(n as u64)
+        .unwrap_or_else(|| panic!("no root of unity of size {} for this field", n));
+    let roots: Vec<Fr<E>> = (0..n).map(|i| omega.pow([i as u64])).collect();
```
Comment (alinush) [inline crates/aptos-dkg/tests/barycentric_eval.rs]: `roots_of_unity`
or
`omegas`
would be clearer: seeing `roots` here could be interpreted as the roots of the polynomial

File: crates/aptos-dkg/tests/barycentric_eval.rs [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+        assert_eq!(
+            val, expected,
+            "Failed for degree {}, n = {} at x = {:?}",
+            degree, n, x
+        );
+    }
+
+    // Interpolation points (roots)
+    for (root, &eval) in roots.iter().zip(evals.iter()) {
+        let val = barycentric_eval(&evals, &roots, *root, n_inv);
```
Comment (alinush) [inline crates/aptos-dkg/tests/barycentric_eval.rs]: This is testing the trivial case where one uses (although they need not to) barycentric interpolation to evaluate at a root of unity, right?

File: crates/aptos-dkg/tests/barycentric_eval.rs [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+            root, degree, n
+        );
+    }
+}
+
+#[test]
+fn test_barycentric_eval_bls12_381() {
+    use ark_bls12_381::Bls12_381;
+    use ark_bn254::Bn254;
+    let cases = [
```
Comment (alinush) [inline crates/aptos-dkg/tests/barycentric_eval.rs:61]: Don't go too high in degree here: your tests will be slow & it won't provide extra coverage it seems. Maybe stop at 16.

Test more edge cases too: play with the size of `n` relative to `degree`, including `n`'s that are too small.

File: crates/aptos-dkg/tests/range_proof.rs [hunk 23]
```diff
@@ -88,34 +80,65 @@ const TEST_CASES: &[(usize, usize)] = &[
...
+    // (255, 16),  Commented out to improve test speed
+    // (255, 32),
+    // (512, 32),
+    // (1024, 32),
+    // (2047, 32),
 ];
 
+#[cfg(test)]
+/// A **reusable** setup structure.
+struct RangeProofFixedSetup<E: Pairing, B: BatchedRangeProof<E>> {
```
Comment (alinush) [inline crates/aptos-dkg/tests/range_proof.rs]: `RangeProofUniversalSetup`?

File: crates/aptos-dkg/tests/range_proof.rs [hunk 24]
```diff
@@ -88,34 +80,65 @@ const TEST_CASES: &[(usize, usize)] = &[
...
+    pk: B::ProverKey,
+    vk: B::VerificationKey,
+}
+
+#[cfg(test)]
+/// Generate a fixed setup for a single curve
+fn make_single_curve_setup<B, E>(n: usize, ell: usize) -> RangeProofFixedSetup<E, B>
```
Comment (alinush) [inline crates/aptos-dkg/tests/range_proof.rs]: Hm... Do you have a function that does set ups for multiple curves? A bit confused: all algorithms are parameterized by the curve via `E`, no?

Comment (alinush) [reply crates/aptos-dkg/tests/range_proof.rs]: Ah, I see the next function below...

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs (hunk 9 above)
Comment (alinush) [reply crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:155]: Yeah, this happens and it's very annoying to deal with... It'd be nice to eventually figure out a design pattern that deals with this automatically in the code & picks the fastest method. TBD.

File: crates/aptos-dkg/src/utils/test_utils.rs (hunk 14 above)
Comment (alinush) [reply crates/aptos-dkg/src/utils/test_utils.rs:15]: Ah! Not sure. Maybe not then!

File: crates/aptos-dkg/tests/barycentric_eval.rs (hunk 22 above)
Comment (waamm) [reply crates/aptos-dkg/tests/barycentric_eval.rs:61]: Ok, they're now ~10x faster on my machine (~0.05s).

File: crates/aptos-dkg/tests/barycentric_eval.rs (hunk 21 above)
Comment (waamm) [reply crates/aptos-dkg/tests/barycentric_eval.rs]: Yep, added some more comments. Could also remove it if it's too trivial.

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs (hunk 18 above)
Comment (waamm) [reply crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:190]: `MsmMap` or `FixedBaseHomomorphism` might be slightly better?

File: crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs (hunk 9 above)
Comment (alinush) [reply crates/aptos-dkg/src/pcs/univariate_hiding_kzg.rs:155]: May wanna leave a `TODO(Perf):`

File: crates/aptos-dkg/tests/barycentric_eval.rs (hunk 21 above)
Comment (alinush) [reply crates/aptos-dkg/tests/barycentric_eval.rs]: Don't remove it cause it's a good thing to test for.

**Review by alinush — APPROVED** [review]

**Comment by alinush** [comment]
Or, @rex1fernando, if you don't have time to review, can you just stamp?

**Review by zjma — APPROVED** [review]
🙈
